lib/StackAnalysis/StackAnalysis.cpp

```diff
@@ -324,6 +354,569 @@ void commitToModel(GeneratedCodeBasicInfo &GCBI,
+  std::set<GlobalVariable *> CalleeSavedRegs;
+  std::set<GlobalVariable *> ClobberedRegs(ABIRegisters.begin(),
+                                           ABIRegisters.end());
+
+  for (CallBase *Call : callers(IndirectBranchInfoMarker)) {
+    auto *CI = cast<CallInst>(Call);
+    if (CI->getParent()->getParent() == F) {
+      if (isa<ConstantInt>(CI->getArgOperand(0))) {
+        if (isa<ConstantInt>(CI->getArgOperand(1))) {
+          int64_t FSO = cast<ConstantInt>(CI->getArgOperand(1))->getSExtValue();
+          if (FSO >= 0)
+            MaybeReturns.emplace_back(CI, FSO);
+          else
+            Result.insert({ CI, FunctionEdgeTypeValue::BrokenReturn });
+        }
+      }
+    }
+  }
+
+  // Elect a final stack offset.
+  const auto WinFSO = electFSO(MaybeReturns);
+
+  // Did we find a valid return instruction?
+  for (const auto &E : MaybeReturns) {
+    auto *CI = E.first;
+    if (E.second == *WinFSO) {
+      Result.insert({ E.first, FunctionEdgeTypeValue::Return });
+      unsigned NumArgs = CI->getNumArgOperands();
+      if (NumArgs > 3) {
+        for (unsigned Idx = 3; Idx < NumArgs; ++Idx) {
+          if (isa<ConstantInt>(CI->getArgOperand(Idx))) {
+            if (cast<ConstantInt>(CI->getArgOperand(Idx))->getZExtValue() == 0)
+              CalleeSavedRegs.insert(ABIRegisters[Idx - 3]);
+          }
+        }
+      }
+    } else {
+      Result.insert({ E.first, FunctionEdgeTypeValue::BrokenReturn });
+    }
+  }
+
+  // Neither a return nor a broken return?
+  for (CallBase *Call : callers(IndirectBranchInfoMarker)) {
+    auto *CI = cast<CallInst>(Call);
+    if (CI->getParent()->getParent() == F) {
+      if (isa<ConstantInt>(CI->getArgOperand(0)))
+        continue;
+
+      if (WinFSO.has_value() && (isa<ConstantInt>(CI->getArgOperand(1)))
+          && (cast<ConstantInt>(CI->getArgOperand(1))->getZExtValue()
+              == *WinFSO))
+        Result.insert({ CI, FunctionEdgeTypeValue::IndirectTailCall });
+      else
+        Result.insert({ CI, FunctionEdgeTypeValue::LongJmp });
+    }
+  }
+
+  bool FoundRet, FoundBrokenRet;
+  FoundRet = FoundBrokenRet = false;
+  for (const auto &E : Result) {
+    if (E.second == FunctionEdgeTypeValue::Return)
+      FoundRet = true;
+    if (E.second == FunctionEdgeTypeValue::BrokenReturn)
+      FoundBrokenRet = true;
+  }
+
+  FunctionTypeValue Type;
+  if (FoundRet) {
+    Type = FunctionTypeValue::Regular;
+  } else if (FoundBrokenRet) {
+    Type = FunctionTypeValue::Fake;
+  } else {
+    Type = FunctionTypeValue::NoReturn;
+  }
+
+  // Finally retrieve the clobbered registers and return them.
+  std::erase_if(ClobberedRegs, [&](const auto &E) {
+    return CalleeSavedRegs.find(E) != CalleeSavedRegs.end();
+  });
+
+  // Refine ABI analyses results.
+  auto *SP = GCBI->spReg();
+  auto It = ABIResults.Arguments.find(SP);
+  if (It != ABIResults.Arguments.end())
+    ABIResults.Arguments.erase(SP);
```

Comment:
Erasing means `Maybe`. We want to say `No`.

Reply:
Isn't `erase` supposed to remove the CSV?

Reply:
Sure. But, as I said, the semantics of missing from the map is `Maybe` (the default). We want `No`.

Reply:
So, instead of erasing we change `RegisterState::Values` of the element to `No`? Makes sense?

Reply:
Yes.